<a href="https://colab.research.google.com/github/MariaNikolaeva2028/LinearRegression/blob/main/LinearRegressionSweets.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Линейная регрессия. Представлены данные, собранные путем голосования за самые популярные конфеты Хэллоуина. Необходимо обучить модель линейной многомерной регрессии. В качестве предикторов выступают поля: chocolate, fruity, caramel, peanutyalmondy, nougat, crispedricewafer, hard, bar, pluribus, sugarpercent, pricepercent, отклик — winpercent.В качестве тренировочного набора данных использовать данные из файла, за иключением следующих конфет: Haribo Gold Bears, Snickers.
Задания:
1.Вывести предсказанное значение winpercent для конфеты Haribo Gold Bears.
2.Вывести предсказанное значение winpercent для конфеты Snickers:
3.Вывести предсказанное значение winpercent для конфеты с параметрами:
[1, 0, 1, 1, 0, 1, 1, 1, 1, 0.355, 0.426]
4.Оценить точность модели.



In [34]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
data = pd.read_csv("https://courses.openedu.ru/assets/courseware/v1/a11f0b9cb0074fdacc112eee71eb0b67/asset-v1:ITMOUniversity+DATANTECH2035+fall_2021_3+type@asset+block/candy-data.csv", delimiter = ',', index_col = 'competitorname')
data

chocolate  fruity  caramel  peanutyalmondy  nougat  \
competitorname                                                              
100 Grand                      1       0        1               0       0   
3 Musketeers                   1       0        0               0       1   
One dime                       0       0        0               0       0   
One quarter                    0       0        0               0       0   
Air Heads                      0       1        0               0       0   
...                          ...     ...      ...             ...     ...   
Snickers Crisper               1       0        1               1       0   
Sour Patch Kids                0       1        0               0       0   
Sour Patch Tricksters          0       1        0               0       0   
Starburst                      0       1        0               0       0   
Strawberry bon bons            0       1        0               0       0   

                       crispedricewafer  hard  bar  pluribus  sugarpercent  \
competitorname                                                               
100 Grand                             1     0    1         0         0.732   
3 Musketeers                          0     0    1         0         0.604   
One dime                              0     0    0         0         0.011   
One quarter                           0     0    0         0         0.011   
Air Heads                             0     0    0         0         0.906   
...                                 ...   ...  ...       ...           ...   
Snickers Crisper                      1     0    1         0         0.604   
Sour Patch Kids                       0     0    0         1         0.069   
Sour Patch Tricksters                 0     0    0         1         0.069   
Starburst                             0     0    0         1         0.151   
Strawberry bon bons                   0     1    0         1         0.569   

                       pricepercent  winpercent  Y  
competitorname                                      
100 Grand                     0.860   66.971725  1  
3 Musketeers                  0.511   67.602936  1  
One dime                      0.116   32.261086  0  
One quarter                   0.511   46.116505  0  
Air Heads                     0.511   52.341465  1  
...                             ...         ... ..  
Snickers Crisper              0.651   59.529251  1  
Sour Patch Kids               0.116   59.863998  1  
Sour Patch Tricksters         0.116   52.825947  1  
Starburst                     0.220   67.037628  1  
Strawberry bon bons           0.058   34.578991  0  

[70 rows x 13 columns]

Создадим тренировочный набор данных:исключив следующие конфеты: Haribo Gold Bears, Snickers.

In [35]:
train_data = data.drop(['Haribo Gold Bears', 'Snickers'])
train_data

chocolate  fruity  caramel  peanutyalmondy  nougat  \
competitorname                                                              
100 Grand                      1       0        1               0       0   
3 Musketeers                   1       0        0               0       1   
One dime                       0       0        0               0       0   
One quarter                    0       0        0               0       0   
Air Heads                      0       1        0               0       0   
...                          ...     ...      ...             ...     ...   
Snickers Crisper               1       0        1               1       0   
Sour Patch Kids                0       1        0               0       0   
Sour Patch Tricksters          0       1        0               0       0   
Starburst                      0       1        0               0       0   
Strawberry bon bons            0       1        0               0       0   

                       crispedricewafer  hard  bar  pluribus  sugarpercent  \
competitorname                                                               
100 Grand                             1     0    1         0         0.732   
3 Musketeers                          0     0    1         0         0.604   
One dime                              0     0    0         0         0.011   
One quarter                           0     0    0         0         0.011   
Air Heads                             0     0    0         0         0.906   
...                                 ...   ...  ...       ...           ...   
Snickers Crisper                      1     0    1         0         0.604   
Sour Patch Kids                       0     0    0         1         0.069   
Sour Patch Tricksters                 0     0    0         1         0.069   
Starburst                             0     0    0         1         0.151   
Strawberry bon bons                   0     1    0         1         0.569   

                       pricepercent  winpercent  Y  
competitorname                                      
100 Grand                     0.860   66.971725  1  
3 Musketeers                  0.511   67.602936  1  
One dime                      0.116   32.261086  0  
One quarter                   0.511   46.116505  0  
Air Heads                     0.511   52.341465  1  
...                             ...         ... ..  
Snickers Crisper              0.651   59.529251  1  
Sour Patch Kids               0.116   59.863998  1  
Sour Patch Tricksters         0.116   52.825947  1  
Starburst                     0.220   67.037628  1  
Strawberry bon bons           0.058   34.578991  0  

[68 rows x 13 columns]

Отбор данных для предикторов:

In [36]:
x = pd.DataFrame(train_data.drop(['winpercent', 'Y'], axis = 1))
x

chocolate  fruity  caramel  peanutyalmondy  nougat  \
competitorname                                                              
100 Grand                      1       0        1               0       0   
3 Musketeers                   1       0        0               0       1   
One dime                       0       0        0               0       0   
One quarter                    0       0        0               0       0   
Air Heads                      0       1        0               0       0   
...                          ...     ...      ...             ...     ...   
Snickers Crisper               1       0        1               1       0   
Sour Patch Kids                0       1        0               0       0   
Sour Patch Tricksters          0       1        0               0       0   
Starburst                      0       1        0               0       0   
Strawberry bon bons            0       1        0               0       0   

                       crispedricewafer  hard  bar  pluribus  sugarpercent  \
competitorname                                                               
100 Grand                             1     0    1         0         0.732   
3 Musketeers                          0     0    1         0         0.604   
One dime                              0     0    0         0         0.011   
One quarter                           0     0    0         0         0.011   
Air Heads                             0     0    0         0         0.906   
...                                 ...   ...  ...       ...           ...   
Snickers Crisper                      1     0    1         0         0.604   
Sour Patch Kids                       0     0    0         1         0.069   
Sour Patch Tricksters                 0     0    0         1         0.069   
Starburst                             0     0    0         1         0.151   
Strawberry bon bons                   0     1    0         1         0.569   

                       pricepercent  
competitorname                       
100 Grand                     0.860  
3 Musketeers                  0.511  
One dime                      0.116  
One quarter                   0.511  
Air Heads                     0.511  
...                             ...  
Snickers Crisper              0.651  
Sour Patch Kids               0.116  
Sour Patch Tricksters         0.116  
Starburst                     0.220  
Strawberry bon bons           0.058  

[68 rows x 11 columns]

Столбец отклика Y:

In [37]:
y = pd.DataFrame(train_data['winpercent'])
y

winpercent
competitorname                   
100 Grand               66.971725
3 Musketeers            67.602936
One dime                32.261086
One quarter             46.116505
Air Heads               52.341465
...                           ...
Snickers Crisper        59.529251
Sour Patch Kids         59.863998
Sour Patch Tricksters   52.825947
Starburst               67.037628
Strawberry bon bons     34.578991

[68 rows x 1 columns]

Обучение модели:

In [38]:
reg = LinearRegression().fit(x, y)

Предсказание для конфет, введенных в ручную:

In [39]:
reg.predict([[1, 0, 1, 1, 0, 1, 1, 1, 1, 0.355, 0.426]])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[60.44836288]])

Предсказание значения для конфеты Haribo Gold Bears:

In [40]:
HariboGoldBears = data.loc['Haribo Gold Bears',:].to_frame().T
reg.predict(HariboGoldBears.drop(['winpercent','Y'], axis=1))

array([[43.76957716]])

Предсказание значения для конфеты Snickers:

In [41]:
Snickers = data.loc['Snickers',:].to_frame().T
reg.predict(Snickers.drop(['winpercent', 'Y'], axis=1))

array([[63.76550665]])

Вывод коэффициентов θ0 и θ1,θ2 ... θp:

In [42]:
reg.intercept_

array([37.06211694])

In [43]:
reg.coef_

array([[25.83115179,  9.85292222, -2.93618458,  7.08036038,  2.09795082,
        10.30185205, -7.25254491, -5.57031332, -4.17224529, 11.78163601,
        -9.5734999 ]])

Оценка точности модели:

In [44]:
y_true = y['winpercent'].to_numpy()
y_new = reg.predict(x)

In [45]:
r2_score(y_true,y_new)

0.5927713036860534

Таким образом, точность предсказанной модели составляет 59%. Очевидно, что зависимость между признаками не является линейной.